### DISCUSSION TITLE
...utils\request.ts   80:19  warning  Forbidden non-null assertion  @typescript-eslint/no-non-null-assertion

### DISCUSSION THREAD

**Discussion by nkamp** [axios/axios#6451] [Q&A]
I'm upgrading Axios from 0.22.0 to 1.7.2. But I get now the warning: 80:19  warning  Forbidden non-null assertion  @typescript-eslint/no-non-null-assertion

```
axiosInstance.interceptors.response.use(

  function (response): AxiosResponse {
    return response;
  },
  function (error: AxiosError) {
    if (
      error.response?.status === HTTP_UNAUTHENTICATED &&
      error.config?.method !== "head"
    ) {
      return handleUnauthenticated(error);
    } else if (
      error.response?.status === HTTP_NOT_FOUND &&
      !isReadOnly(error.config!)                                   // This line without ! gives an error
    ) {
      router.currentRoute.value.meta.ignoreChanges = true;
      router.push({ name: "home" }).catch(ignoreException).then();
    }

    if (
      error.response?.headers["content-type"] ===
        "application/problem+json; charset=utf-8" &&
      error.config?.headers &&
      error.config?.headers[ERROR_HANDLING_HEADER] !== LOCAL_ERROR_HANDLING
    ) {
      if (error.response?.status === HTTP_PRECONDITION_FAILED) {
        createPreconditionFailedNotification(error.response.data as ApiError);
      } else {
        createErrorNotification(error.response.data as ApiError);
      }
    }
    return Promise.reject(error);
  }
);
```

I have added the '!' to the  line `!isReadOnly(error.config!)`

I don't know what this exactly does except that I say now 'it will be never undefined', but what is the consequences if I do this? 
And I'm asking myself and I don't believe it, that this is the right way to solve this. My application is working without 'issue'.
And off course I can add the rule ' @typescript-eslint/no-non-null-assertion', but I don't think it is a good idea.

Can someone explain what is the right way to solve such a issue?

**Comment by jasonsaayman** [comment]
`error.config!` is a TypeScript non-null assertion. It does **not** add any runtime check. It only tells TypeScript: “trust me, this value is not `undefined`.”

So this:

```ts
isReadOnly(error.config!)
```

compiles as if `error.config` is always present. If it is actually `undefined` at runtime, then `isReadOnly()` receives `undefined`, which may cause a runtime error depending on how `isReadOnly()` is implemented.

In Axios 1.x, `AxiosError.config` is typed as optional:

```ts
config?: InternalAxiosRequestConfig<D>;
```

That is intentional because some Axios errors can happen before or outside a fully available request config.

The better fix is to narrow the value before using it:

```ts
const response = error.response;
const config = error.config;

if (
  response?.status === HTTP_NOT_FOUND &&
  config &&
  !isReadOnly(config)
) {
  router.currentRoute.value.meta.ignoreChanges = true;
  router.push({ name: 'home' }).catch(ignoreException).then();
}
```

Another good option is to make `isReadOnly()` accept an optional config if `undefined` has a clear meaning in your app:

```ts
function isReadOnly(config?: AxiosError['config']) {
  return config?.method === 'get' || config?.method === 'head';
}
```

Then you can call it without `!`:

```ts
!isReadOnly(error.config)
```

If the branch is specifically about an HTTP response, you can also use `error.response.config`, because `AxiosResponse.config` is not optional once a response exists:

```ts
const response = error.response;

if (
  response?.status === HTTP_NOT_FOUND &&
  !isReadOnly(response.config)
) {
  router.currentRoute.value.meta.ignoreChanges = true;
  router.push({ name: 'home' }).catch(ignoreException).then();
}
```

So the short answer is: avoid `error.config!` unless you are absolutely sure it cannot be missing. Prefer a guard, optional parameter handling, or `error.response.config` when you already know a response exists.
